spectrum-network/src/peer_manager.rs

```diff
@@ -184,67 +199,187 @@ impl PeerEvents for PeersMailbox {
+            trace!("Going to connect peer {}", pid);
+            self.connect(&pid)
         }
     }
 
-    fn connect(&mut self, peer_id: PeerId) {
-        if let Some(PeerInState::NotConnected(ncp)) = self.state.peer(&peer_id) {
-            let should_connect = if let Some(backoff_until) = ncp.backoff_until() {
-                backoff_until <= Instant::now()
-            } else {
-                true
-            };
-            if should_connect && ncp.try_connect().is_ok() {
-                self.out_queue.push_back(PeerManagerOut::Connect(peer_id))
+    /// Connect to a known peer.
+    fn connect(&mut self, peer_id: &PeerId) {
+        trace!("Connect(peer_id={})", peer_id);
+        if let Some(PeerInState::NotConnected(ncp)) = self.state.peer(peer_id) {
+            if ncp
+                .backoff_until()
+                .map(|backoff_until| backoff_until <= Instant::now())
+                .unwrap_or(true)
+            {
+                let cp = ncp.connect();
+                self.out_queue
+                    .push_back(PeerManagerOut::Connect(cp.destination()))
+            }
+        }
+    }
+
+    /// Disconnect a known peer.
+    fn disconnect(&mut self, peer_id: PeerId, forget: bool) {
+        if let Some(PeerInState::Connected(cp)) = self.state.peer(&peer_id) {
+            let ncp = cp.disconnect();
+            trace!("Peer {} disconnected", peer_id);
+            if forget {
+                ncp.forget();
+                trace!("Peer {} forgotten", peer_id);
+            }
+            self.out_queue.push_back(PeerManagerOut::Drop(peer_id));
+        } else {
+            error!("Cannot disconnect peer {}", peer_id);
+        }
+    }
+
+    /// Prepare for new conn allocation cycle.
+    fn prepare_allocate_connections(&mut self) -> ConnAllocationMode {
+        match self.state.networking_state() {
+            NetworkingState::NotBootstrapped(boot_peers) => {
+                let mut added = 0;
+                for p in boot_peers.into_iter() {
+                    if self.state.try_add_peer(p, false, true).is_some() {
+                        added += 1;
+                    }
+                }
+                trace!("Node not bootstrapped. {} peers added.", added);
+                ConnAllocationMode::Active
+            }
+            NetworkingState::BootInProgress => {
+                self.boot_in_progress = true;
+                ConnAllocationMode::Active
+            }
+            NetworkingState::Bootstrapped => {
+                if self.boot_in_progress {
+                    for pid in self
+                        .state
+                        .filter_peers(|_: &PeerId, pif: &PeerInfo| pif.is_boot && pif.state.is_connected())
+                        .into_iter()
+                    {
+                        self.disconnect(pid, true);
+                    }
+                    self.boot_in_progress = false;
+                }
+                ConnAllocationMode::Passive
```

Comment:
If we're already in `Bootstrapped` state, why do we disconnect from peers in this manner? Isn't it possible to disconnect enough peers to put us back into `BootInProgress` state?

Reply:
when boot state = Bootstrapped and boot_in_progress = true the node has just finished bootstrapping. Once the node gained enough outbound connections the best option in terms of security is to get rid of bootstrapping nodes. Usually we use some fixed set of nodes for bootstrapping, so they can potentially be compromised or overloaded with tons of conn requests.

Reply:
It's impossible to get back to BootInProgress as the condition that calculates boot state does not account for boot peers.
https://github.com/spectrum-finance/spectrum/blob/ddb7973802887cefecadc64fdbe9e78754e5ac6b/spectrum-network/src/peer_manager/peers_state.rs#L465

Reply:
Got it, thanks.